java/dataset/src/main/cpp/jni_wrapper.cc

```diff
@@ -126,20 +160,16 @@ class ReserveFromJava : public arrow::dataset::jni::ReservationListener {
       : vm_(vm), java_reservation_listener_(java_reservation_listener) {}
 
   arrow::Status OnReservation(int64_t size) override {
-    JNIEnv* env;
-    if (vm_->GetEnv(reinterpret_cast<void**>(&env), JNI_VERSION) != JNI_OK) {
-      return arrow::Status::Invalid("JNIEnv was not attached to current thread");
-    }
+    JNIEnvGuard guard(vm_);
+    JNIEnv* env = guard.env();
```

Comment:
This function wasn't throwing previously. It was returning an error status previously. Are the callers ready to handle exceptions?

Reply:
try/catch added to maintain the initial behavior